src/client/mod.rs

```diff
@@ -1,322 +1,498 @@
+/// [`Client`]: ./enum.Client.html
 ///
 /// # Examples
 ///
 /// ```rust
-/// use std::env::var_os;
-/// use warpgrapher::client::Client;;
+/// # use warpgrapher::Client;;
 ///
-/// let mut client = Client::new("http://localhost:5000/graphql");
+/// let client = Client::<(), ()>::new_with_http("http://localhost:5000/graphql");
 /// ```
-#[derive(Clone, Hash, Debug, Default)]
-pub struct Client {
-    endpoint: String,
+#[derive(Clone, Debug)]
+pub enum Client<GlobalCtx = (), RequestCtx = ()>
+where
+    GlobalCtx: GlobalContext,
+    RequestCtx: RequestContext,
+{
+    Http {
+        endpoint: String,
+    },
+    Local {
+        engine: Box<Engine<GlobalCtx, RequestCtx>>,
+    },
 }
 
-impl Client {
-    /// Takes an endpoint string and creates a new [`Client`].
+impl<GlobalCtx, RequestCtx> Client<GlobalCtx, RequestCtx>
+where
+    GlobalCtx: GlobalContext,
+    RequestCtx: RequestContext,
+{
+    /// Takes the URL of a Warpgrapher service endpoint and returns a new ['Client'] initialized to
+    /// query that endpoint.  The type parameters are only relevant for a local instance of the
+    /// Warpgrapher engine, not for a remote HTTP client, so pass () for both type parameters, as
+    /// shown in the example below.
+    ///
+    /// [`Client`]: ./enum.Client.html
     ///
-    /// [`Client`]: ./struct.Client.html
     ///
     /// # Examples
     ///
     /// ```rust
-    /// use std::env::var_os;
-    /// use warpgrapher::client::Client;
+    /// # use warpgrapher::Client;
     ///
-    /// let mut client = Client::new("http://localhost:5000/graphql");
+    /// let mut client = Client::<(), ()>::new_with_http("http://localhost:5000/graphql");
     /// ```
-    pub fn new(endpoint: &str) -> Client {
-        Client {
+    pub fn new_with_http(endpoint: &str) -> Client<(), ()> {
+        trace!("Client::new_with_http called -- endpoint: {}", endpoint);
+
+        Client::Http {
             endpoint: endpoint.to_string(),
         }
     }
 
-    /// Executes a raw graphql query.
+    /// Takes a Warpgrapher engine and returns a new ['Client'] initialized to query that engine.
+    /// The type parameters are the [`GlobalContext`] and [`RequestContext`] used by the engine.
+    ///
+    /// [`Client`]: ./enum.Client.html
+    /// [`GlobalContext`]: ../engine/context/trait.GlobalContext.html
+    /// [`RequestContext`]: ../engine/context/trait.RequestContext.html
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use warpgrapher::{Client, Configuration, DatabasePool, Engine};
+    ///
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// let c = Configuration::new(1, Vec::new(), Vec::new());
+    /// let engine = Engine::new(c, DatabasePool::NoDatabase).build()?;
+    ///
+    /// let mut client = Client::<(), ()>::new_with_local(engine);
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn new_with_local(engine: Engine<GlobalCtx, RequestCtx>) -> Client<GlobalCtx, RequestCtx> {
```

Comment:
@maldrake : how would you feel about renaming this to `new_with_engine` ?

Reply:
Agreed.  Changed.